core/src/main/java/org/infinispan/container/AbstractSegmentedDataContainer.java

```diff
@@ -0,0 +1,634 @@
+      return Spliterators.spliteratorUnknownSize(iterator(), Spliterator.DISTINCT | Spliterator.CONCURRENT | Spliterator.NONNULL);
+   }
+
+   @Override
+   public Spliterator<InternalCacheEntry<K, V>> spliteratorIncludingExpired(IntSet segments) {
+      return Spliterators.spliteratorUnknownSize(iteratorIncludingExpired(segments), Spliterator.DISTINCT | Spliterator.CONCURRENT | Spliterator.NONNULL);
+   }
+
+   @Override
+   public Spliterator<InternalCacheEntry<K, V>> spliteratorIncludingExpired() {
+      return Spliterators.spliteratorUnknownSize(iteratorIncludingExpired(), Spliterator.DISTINCT | Spliterator.CONCURRENT | Spliterator.NONNULL);
+   }
+
+   @Override
+   public void clear(IntSet segments) {
+      Iterator<InternalCacheEntry<K, V>> iter = iteratorIncludingExpired(segments);
+      while (iter.hasNext()) {
+         iter.next();
+         iter.remove();
+      }
+   }
+
+   @Override
+   public void clear() {
+      Iterator<InternalCacheEntry<K, V>> iter = iteratorIncludingExpired();
+      while (iter.hasNext()) {
+         iter.next();
+         iter.remove();
+      }
+   }
+
+   /**
+    * This method is invoked every time an entry is written inside a compute block
+    * @param key key passed to compute method
+    * @param value the new value
+    */
+   protected void computeEntryWritten(K key, InternalCacheEntry<K, V> value) {
+      // Do nothing by default
+   }
+
+   /**
+    * This method is invoked every time an entry is removed inside a compute block
+    * @param key key passed to compute method
+    * @param value the old value
+    */
+   protected void computeEntryRemoved(K key, InternalCacheEntry<K, V> value) {
+      // Do nothing by default
+   }
+
+   @Override
+   public void addRemovalListener(Consumer<Iterable<InternalCacheEntry<K, V>>> listener) {
+      listeners.add(listener);
+   }
+
+   @Override
+   public void removeRemovalListener(Object listener) {
+      listeners.remove(listener);
+   }
+
+   protected class EntryIterator extends AbstractIterator<InternalCacheEntry<K, V>> {
+
+      private final Iterator<InternalCacheEntry<K, V>> it;
+
+      public EntryIterator(Iterator<InternalCacheEntry<K, V>> it) {
+         this.it = it;
+      }
+
+      protected InternalCacheEntry<K, V> getNext() {
+         boolean initializedTime = false;
+         long now = 0;
+         while (it.hasNext()) {
+            InternalCacheEntry<K, V> entry = it.next();
+            if (!entry.canExpire()) {
+               if (trace()) {
+                  log().tracef("Return next entry %s", entry);
+               }
+               return entry;
+            } else {
+               if (!initializedTime) {
+                  now = timeService.wallClockTime();
+                  initializedTime = true;
+               }
+               if (!entry.isExpired(now) || expirationManager.entryExpiredInMemoryFromIteration(entry, now).join()
+                     == Boolean.FALSE) {
```

Comment:
The idea was if we allowed for some sort of non blocking variant for iteration (which we don't have a non blocking iteration yet). We can discuss this in a different JIRA maybe?

Reply:
Sure, I just want to make sure we don't implement something "just in case" and when we get to actually using it we get stuck doing it the same way because "we've always done it this way"